core/shared/src/main/scala/fs2/Stream.scala

```diff
@@ -1984,6 +1984,42 @@ object Stream {
       self.reduce(S.combine(_, _))
 
     /**
+      * Repartitions the input with the function `p`. On each step `p` is applied
+      * to the input and all elements but the last of the resulting sequence
+      * are emitted. The last element is then appended to the next input using the
+      * Semigroup `S`.
+      *
+      * @example {{{
+      * scala> import cats.implicits._
+      * scala> Stream("Hel", "l", "o Wor", "ld").repartition(s => Segment.array(s.split(" "))).toList
+      * res0: List[String] = List(Hello, World)
+      * }}}
+      */
+    def repartition(p: O => Segment[O, Unit])(implicit S: Semigroup[O]): Stream[F,O] = {
+      def go(carry: Option[O], s: Stream[F,O]): Pull[F,O,Unit] = {
+        s.pull.uncons.flatMap {
+          case Some((ohd, otl)) =>
+            ohd.uncons1 match {
```

Comment:
Can you try using `Segment#scan` or `Segment#fold` here instead of doing an `uncons1` and pushing the remainder back in to `s`? We've hit a lot of performance issues due to patterns resembling an uncons followed by a push back.